### PR TITLE
Fix execution issues in distributed_training_with_torch guide

### LINKED ISSUE #23453: [Fix] Resolve PyTorch DDP execution crashes, NCHW shape errors, and sampler state
### Description of the Bug
An evaluation of the *Multi-GPU distributed training with PyTorch* guide revealed several critical execution and setup issues. When executed as provided, the script fails due to data shape incompatibilities, multiprocessing deadlocks, loss function target type mismatches, and improper distributed dataset sampling.

---

### Key Issues Identified

1. **Multiprocessing Start Method Deadlock (`fork` vs `spawn`)**
   * **Location:** Execution block (`torch.multiprocessing.start_processes`)
   * **Problem:** The script invokes `torch.cuda.device_count()` at the top level before calling `start_processes(..., start_method="fork")`. In PyTorch, using `fork` after initializing the CUDA context leads to child process deadlocks and runtime failures.
   * **Fix:** Change `start_method` to `"spawn"`.

2. **Tensor Shape Incompatibility (`NHWC` vs `NCHW`)**
   * **Location:** `get_dataset()`
   * **Problem:** Keras data loading outputs images in `NHWC` format `(28, 28, 1)`. PyTorch convolution layers (`Conv2D`) expect `NCHW` layout `(1, 28, 28)`. Passing `NHWC` arrays causes dimension errors during the forward pass.
   * **Fix:** Transpose array dimensions via `np.transpose(x_train, (0, 3, 1, 2))` before generating the PyTorch `TensorDataset`.

3. **Target Label Data Type Mismatch**
   * **Location:** `get_dataset()` / `train_model()`
   * **Problem:** `y_train` is loaded as an integer array, but PyTorch's `nn.CrossEntropyLoss` specifically requires target indices to be explicitly typed as 64-bit integers (`torch.int64` / `LongTensor`).
   * **Fix:** Cast labels using `y_train.astype("int64")`.

4. **Missing `set_epoch` in `DistributedSampler`**
   * **Location:** `train_model()`
   * **Problem:** `DistributedSampler.set_epoch(epoch)` is omitted in the training loop. Without this, the dataset ordering remains deterministic across epochs, preventing effective data shuffling across GPU replicas.
   * **Fix:** Invoke `sampler.set_epoch(epoch)` at the beginning of each epoch iteration.

5. **Unhandled Exception in Zero-GPU Environments**
   * **Location:** Top-level runtime execution
   * **Problem:** If executed in an environment without active CUDA GPUs, `num_gpu` evaluates to `0`, causing `nprocs=0` in `start_processes()` to throw an uncaught `ValueError`.
   * **Fix:** Add a validation guard to verify `num_gpu > 0`.

---

### Proposed Fix / Corrected Implementation

```python
import os
import numpy as np
import torch
import keras

def get_model():
    inputs = keras.Input(shape=(28, 28, 1))
    x = keras.layers.Rescaling(1.0 / 255.0)(inputs)
    x = keras.layers.Conv2D(filters=12, kernel_size=3, padding="same", use_bias=False)(x)
    x = keras.layers.BatchNormalization(scale=False, center=True)(x)
    x = keras.layers.ReLU()(x)
    x = keras.layers.Conv2D(filters=24, kernel_size=6, use_bias=False, strides=2)(x)
    x = keras.layers.BatchNormalization(scale=False, center=True)(x)
    x = keras.layers.ReLU()(x)
    x = keras.layers.Conv2D(filters=32, kernel_size=6, padding="same", strides=2, name="large_k")(x)
    x = keras.layers.BatchNormalization(scale=False, center=True)(x)
    x = keras.layers.ReLU()(x)
    x = keras.layers.GlobalAveragePooling2D()(x)
    x = keras.layers.Dense(256, activation="relu")(x)
    x = keras.layers.Dropout(0.5)(x)
    outputs = keras.layers.Dense(10)(x)
    return keras.Model(inputs, outputs)

def get_dataset():
    (x_train, y_train), _ = keras.datasets.mnist.load_data()

    # Convert layout from NHWC (28, 28, 1) to NCHW (1, 28, 28) for PyTorch
    x_train = x_train.astype("float32")
    x_train = np.expand_dims(x_train, -1)
    x_train = np.transpose(x_train, (0, 3, 1, 2))

    # Ensure label targets match PyTorch int64 requirements
    y_train = y_train.astype("int64")

    return torch.utils.data.TensorDataset(
        torch.from_numpy(x_train), torch.from_numpy(y_train)
    )

def setup_device(current_gpu_index, num_gpus):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "56492"
    torch.cuda.set_device(current_gpu_index)
    torch.distributed.init_process_group(
        backend="nccl",
        init_method="env://",
        world_size=num_gpus,
        rank=current_gpu_index,
    )

def cleanup():
    torch.distributed.destroy_process_group()

def prepare_dataloader(dataset, current_gpu_index, num_gpus, batch_size):
    sampler = torch.utils.data.distributed.DistributedSampler(
        dataset,
        num_replicas=num_gpus,
        rank=current_gpu_index,
        shuffle=True,
    )
    dataloader = torch.utils.data.DataLoader(
        dataset,
        sampler=sampler,
        batch_size=batch_size,
        shuffle=False,
        num_workers=2,
        pin_memory=True,
    )
    return dataloader, sampler

def train_model(model, dataloader, sampler, num_epochs, optimizer, loss_fn, current_gpu_index):
    for epoch in range(num_epochs):
        # Synchronize sampler random seed per epoch
        sampler.set_epoch(epoch)
        running_loss = 0.0
        running_loss_count = 0

        for batch_idx, (inputs, targets) in enumerate(dataloader):
            inputs = inputs.to(current_gpu_index, non_blocking=True)
            targets = targets.to(current_gpu_index, non_blocking=True)

            outputs = model(inputs)
            loss = loss_fn(outputs, targets)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            running_loss += loss.item()
            running_loss_count += 1

        if current_gpu_index == 0:
            print(f"Epoch {epoch + 1}/{num_epochs} - Loss: {running_loss / running_loss_count:.4f}")

def per_device_launch_fn(current_gpu_index, num_gpu, num_epochs, batch_size):
    setup_device(current_gpu_index, num_gpu)

    dataset = get_dataset()
    model = get_model()
    dataloader, sampler = prepare_dataloader(dataset, current_gpu_index, num_gpu, batch_size)

    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    loss_fn = torch.nn.CrossEntropyLoss()

    model = model.to(current_gpu_index)
    ddp_model = torch.nn.parallel.DistributedDataParallel(
        model, device_ids=[current_gpu_index], output_device=current_gpu_index
    )

    train_model(ddp_model, dataloader, sampler, num_epochs, optimizer, loss_fn, current_gpu_index)
    cleanup()

if __name__ == "__main__":
    num_gpu = torch.cuda.device_count()
    num_epochs = 2
    batch_size = 64

    if num_gpu < 1:
        raise RuntimeError("No CUDA GPUs detected. Multi-GPU training requires at least 1 GPU.")

    torch.multiprocessing.start_processes(
        per_device_launch_fn,
        args=(num_gpu, num_epochs, batch_size),
        nprocs=num_gpu,
        join=True,
        start_method="spawn",
    )

### PR DESCRIPTION
**Pull request by HEMANTHHP2005**
Fixes #23453

### Summary of changes:
- Switched multiprocessing start method from `fork` to `spawn` to avoid CUDA context deadlocks.
- Added channel transposition to convert image inputs to PyTorch's expected NCHW layout.
- Cast target labels to `int64` for compatibility with `nn.CrossEntropyLoss`.
- Added `dataloader.sampler.set_epoch(epoch)` inside the epoch training loop.
- Added a guard for environments with 0 CUDA devices.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Comment by HEMANTHHP2005** [comment]
Hi @gbaned, 

All PyTorch DDP fixes are in place:
- Switched multiprocessing to `spawn` method to prevent CUDA context deadlocks.
- Transposed inputs to NCHW format.
- Cast target labels to `int64`.
- Added distributed sampler epoch tracking and rank-0 print guards.

All framework unit tests and `ruff` format checks are passing. The failure on `api_gen` in the format workflow is due to autogenerated root API files, which does not affect this guide. Ready for review!

**Comment by hertschuh** [comment]
This file was deleted. These guides where outdated versions of the guides on keras.io.

This one should be modified instead: https://github.com/keras-team/keras-io/blob/master/guides/distributed_training_with_torch.py
